core/util/src/main/java/org/eclipse/rdf4j/common/iteration/TimeLimitIteration.java

```diff
@@ -78,23 +79,41 @@ public void remove()
 	protected void handleClose()
 		throws X
 	{
-		interruptTask.cancel();
```

Comment:
This seems to have something to do with it, ie, reordering handleClose and interruptTask.cancel may be causing issues. The livelock for me is inside of the infamous testOrderByQueriesAreInterruptable on line 2095 (the hasNext call and it is livelocked trying to complete TimeLimitIteration.interrupt, which is calling TImeLimitIteration.handleClose, eventually getting stuck at DelayedIteration.handleClose:

```
  [1] org.eclipse.rdf4j.common.iteration.DelayedIteration.handleClose (DelayedIteration.java:118), pc = 8
  [2] org.eclipse.rdf4j.query.algebra.evaluation.iterator.OrderIterator.handleClose (OrderIterator.java:232), pc = 1
  [3] org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close (AbstractCloseableIteration.java:49), pc = 13
  [4] org.eclipse.rdf4j.common.iteration.Iterations.closeCloseable (Iterations.java:121), pc = 11
  [5] org.eclipse.rdf4j.common.iteration.IterationWrapper.handleClose (IterationWrapper.java:99), pc = 8
  [6] org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close (AbstractCloseableIteration.java:49), pc = 13
  [7] org.eclipse.rdf4j.common.iteration.Iterations.closeCloseable (Iterations.java:121), pc = 11
  [8] org.eclipse.rdf4j.common.iteration.IterationWrapper.handleClose (IterationWrapper.java:99), pc = 8
  [9] org.eclipse.rdf4j.sail.base.SailClosingIteration.handleClose (SailClosingIteration.java:124), pc = 1
  [10] org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close (AbstractCloseableIteration.java:49), pc = 13
  [11] org.eclipse.rdf4j.common.iteration.Iterations.closeCloseable (Iterations.java:121), pc = 11
  [12] org.eclipse.rdf4j.common.iteration.IterationWrapper.handleClose (IterationWrapper.java:99), pc = 8
  [13] org.eclipse.rdf4j.sail.helpers.SailBaseIteration.handleClose (SailBaseIteration.java:62), pc = 1
  [14] org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close (AbstractCloseableIteration.java:49), pc = 13
  [15] org.eclipse.rdf4j.common.iteration.Iterations.closeCloseable (Iterations.java:121), pc = 11
  [16] org.eclipse.rdf4j.common.iteration.IterationWrapper.handleClose (IterationWrapper.java:99), pc = 8
  [17] org.eclipse.rdf4j.common.iteration.TimeLimitIteration.handleClose (TimeLimitIteration.java:83), pc = 1
  [18] org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close (AbstractCloseableIteration.java:49), pc = 13
  [19] org.eclipse.rdf4j.common.iteration.TimeLimitIteration.interrupt (TimeLimitIteration.java:119), pc = 16
  [20] org.eclipse.rdf4j.common.iteration.InterruptTask.run (InterruptTask.java:30), pc = 16
  [21] java.util.TimerThread.mainLoop (Timer.java:555), pc = 221
  [22] java.util.TimerThread.run (Timer.java:505), pc = 1
```